include/patchworkpp/patchworkpp.hpp

```diff
@@ -1021,4 +1040,134 @@ void PatchWorkpp<PointT>::pc2czm(const pcl::PointCloud<PointT> &src, std::vector
     if (verbose_) cout << "[ CZM ] Divides pointcloud into the concentric zone model" << endl;
 }
 
+template<typename PointT> inline
+rcl_interfaces::msg::SetParametersResult
+PatchWorkpp<PointT>::onSetParameters(const std::vector<rclcpp::Parameter> &params) const
+{
+    rcl_interfaces::msg::SetParametersResult result;
+    result.successful = true;
+    result.reason = "Parameters validated successfully.";
+
+    // Create temporary copies to validate parameter changes
+    int tmp_num_zones = num_zones_;
+    std::vector<long> tmp_num_sectors = num_sectors_each_zone_;
+    std::vector<long> tmp_num_rings = num_rings_each_zone_;
+    std::vector<double> tmp_elevation_thr = elevation_thr_;
+    std::vector<double> tmp_flatness_thr = flatness_thr_;
+
+    // Apply incoming parameter changes to our temporary copies
+    for (const auto& param : params) {
+        if (param.get_name() == "czm.num_zones") {
+            tmp_num_zones = param.as_int();
+        }
+        else if (param.get_name() == "czm.num_sectors_each_zone") {
+            tmp_num_sectors = param.as_integer_array();
+        }
+        else if (param.get_name() == "czm.num_rings_each_zone") {
+            tmp_num_rings = param.as_integer_array();
+        }
+        else if (param.get_name() == "czm.elevation_thresholds") {
+            tmp_elevation_thr = param.as_double_array();
+        }
+        else if (param.get_name() == "czm.flatness_thresholds") {
+            tmp_flatness_thr = param.as_double_array();
+        }
+    }
+
+    // Validate parameters
+    if (tmp_num_zones != 4) {
+        result.successful = false;
+        result.reason = "num_zones must be 4";
+    }
+
+    if (tmp_num_sectors.size() != tmp_num_rings.size()) {
+        result.successful = false;
+        result.reason = "num_sectors_each_zone and num_rings_each_zone must have the same size";
+    }
+
+    if (tmp_num_sectors.size() != static_cast<size_t>(tmp_num_zones)) {
+        result.successful = false;
+        result.reason = "num_sectors_each_zone size must match num_zones";
+    }
+
+    if (tmp_elevation_thr.size() != tmp_flatness_thr.size()) {
+        result.successful = false;
+        result.reason = "elevation_thresholds and flatness_thresholds must have the same size";
+    }
+
+    if (!result.successful)
+    {
+        RCLCPP_ERROR_STREAM(rclcpp::get_logger("patchworkpp"), "Parameter validation failed: " << result.reason);
```

Comment:
I think it may be better to print it for each check. rn, if you change num_zones, you still get the error message from `tmp_num_sectors`